### PR TITLE
Descriptor buffer sample comply to vulkan specs

### PR DESCRIPTION
**Pull request by VladyslavSan**
Comply with Vulkan specs about how descriptor bindings are placed into the set layout's memory.

## Description

Please include a summary of the change, new sample, or fixed issue. Please also include relevant motivation and context.
Please read the [contribution guidelines](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md)

Fixes #<issue number>

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
  
## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](./../samples/extensions/conditional_rendering)

### CODE REVIEW

**Review by asuessenbach — DISMISSED** [review]
LGTM.
To express the connection of data, it might be worth to introduce some data struct, like 
```
	struct DescriptorData
	{
		VkDescriptorSetLayout              layout{VK_NULL_HANDLE};
		std::unique_ptr<vkb::core::Buffer> buffer;
		VkDeviceSize                       size;
		VkDeviceSize                       offset;
	};
```
and two members of the DescriptorBufferBasic class like
```
	DescriptorData resource_descriptor;
	DescriptorData image_descriptor;
```

**Comment by VladyslavSan** [comment]
> LGTM. To express the connection of data, it might be worth to introduce some data struct, like
> 
> ```
> 	struct DescriptorData
> 	{
> 		VkDescriptorSetLayout              layout{VK_NULL_HANDLE};
> 		std::unique_ptr<vkb::core::Buffer> buffer;
> 		VkDeviceSize                       size;
> 		VkDeviceSize                       offset;
> 	};
> ```
> 
> and two members of the DescriptorBufferBasic class like
> 
> ```
> 	DescriptorData resource_descriptor;
> 	DescriptorData image_descriptor;
> ```

Agree, that will improve understanding even better hinting that all of those fields are logically connected and used together.

File: samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.h [hunk 1]
```diff
@@ -30,16 +30,17 @@ class DescriptorBufferBasic : public ApiVulkanSample
...
-	VkDescriptorSetLayout descriptor_set_layout_image{VK_NULL_HANDLE};
+	struct DescriptorData
+	{
+		VkDescriptorSetLayout              layout{VK_NULL_HANDLE};
+		std::unique_ptr<vkb::core::Buffer> buffer;
+		VkDeviceSize                       size;
+		VkDeviceSize                       offset;
+	};
+	DescriptorData uniform_binging_descriptor;
+	DescriptorData image_binging_descriptor;
```
Comment (asuessenbach) [inline samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.h]: What does `binging` here mean?

Comment (VladyslavSan) [reply samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.h]: Telling that it's related to set binding. Also without `binding` it sometimes collides with other variables in the code.

Comment (asuessenbach) [reply samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.h]: That is, you probably wanted to use `binding`, not `binging`?

**Review by SaschaWillems — CHANGES_REQUESTED** [review]
Thank you very much for this PR.  Really appreciated.

One thing though: The sample also comes with a tutorial that shows parts of the code, so that needs to be changed too.

**Comment by VladyslavSan** [comment]
> Thank you very much for this PR. Really appreciated.
> 
> One thing though: The sample also comes with a tutorial that shows parts of the code, so that needs to be changed too.

Oh, you mean this one https://github.com/KhronosGroup/Vulkan-Samples/blob/1048740b2a4adde1b8cb01c63732bdde3d42147b/samples/extensions/descriptor_buffer_basic/README.adoc. 
Okay, I'll modify it too.

File: samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.cpp [hunk 2]
```diff
@@ -196,6 +201,10 @@ void DescriptorBufferBasic::setup_descriptor_set_layout()
 	vkGetDescriptorSetLayoutSizeEXT(get_device().get_handle(), uniform_binding_descriptor.layout, &uniform_binding_descriptor.size);
 	vkGetDescriptorSetLayoutSizeEXT(get_device().get_handle(), image_binding_descriptor.layout, &image_binding_descriptor.size);
 
+	// Adjust set layout sizes to alignment.
+	uniform_binding_descriptor.size = aligned_size(uniform_binding_descriptor.size, descriptor_buffer_properties.descriptorBufferOffsetAlignment);
```
Comment (VladyslavSan) [inline samples/extensions/descriptor_buffer_basic/descriptor_buffer_basic.cpp:205]: Just a comment to explain. Before this PR we had this code:
```
uniform_descriptor_offset = aligned_size(descriptor_buffer_properties.uniformBufferDescriptorSize, descriptor_buffer_properties.descriptorBufferOffsetAlignment);
	image_descriptor_offset   = aligned_size(descriptor_buffer_properties.combinedImageSamplerDescriptorSize, descriptor_buffer_properties.descriptorBufferOffsetAlignment);
```
And it was adjusting only the descriptor binding size. Then I removed it, as for the offset calculation entire set layout size should be used. But then while reading adoc and `descriptorBufferOffsetAlignment` I understood that set layout size should still be adjusted to comply with an alignment requirement.

**Review by asuessenbach — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]
LGTM. Thank you very much for this PR!

**Comment by VladyslavSan** [comment]
@SaschaWillems do you know who else can I add as a reviewer to this PR? It says 3 approvals are needed. I don't see in docs anything about some github group or a list of individuals to add to PR.

**Comment by SaschaWillems** [comment]
There is no need to add reviewers. We'll look at this issue in our next samples call.

**Pull request merged**